### ISSUE TITLE
`module.hot` is replaced by `false` on development

### ISSUE THREAD

**Issue by JounQin** [webpack-contrib/webpack-hot-client#5]
This issue is moved from https://github.com/webpack/webpack/issues/6276

* Operating System: macOS 10.12
* Node Version: v9.3.0
* NPM Version: 5.6.0
* webpack Version: 4.0.0-alpha.4

- [x] This is a **bug**
- [ ] This is a **feature** request
- [ ] This is a **modification** request

### Code

```js
// webpack.config.js
import path from 'path'

import HtmlWebpackPlugin from 'html-webpack-plugin'
import webpack from 'webpack'

const NODE_ENV = process.env.NODE_ENV || 'development'

const isProd = NODE_ENV === 'production'

export default {
  mode: NODE_ENV,
  output: {
    filename: `[name].[${isProd ? 'chunkhash' : 'hash'}].js`,
  },
  plugins: [
    new HtmlWebpackPlugin({
      template: 'src/index.html',
      minify: {
        collapseWhitespace: isProd,
      },
    }),
  ],
}
```

```js
// src/index.js
if (module.hot) {
  console.log('This will never be called')
}
```

### Expected Behavior
log 'This will never be called'

### Actual Behavior
`module.hot` is replaced by `false`, and log will never be called

### For Bugs; How can we reproduce the behavior?
https://github.com/JounQin/webpack-next-study/tree/module.hot

branch `module.hot`, simply run `yarn dev` and open http://localhost:4000

**Comment by JounQin** [comment]
I notice you're using `typeof module.hot === 'object'` instead of `module.hot` to check whether enable `module.hot.accept`, and I change into this code in my demo app, that works, so what makes this change? And if it is as expected, then it should be documented.

**Comment by shellscape** [comment]
@JounQin great observation (and I'd like to thank you very much for your interest and activity in these repos - I really appreciate it!).

I ran into a browser quirk that I thought was isolated to my machine, whereby `module.hot` was an object, it had properties, but was not being evaluated by the browser as truthy for some reason. From your issue it looks like it wasn't isolated to my environment.

@sokra please see the message above. What would cause `module.hot` to be falsy, even though it's a valid object with properties? `if (module.hot)` doesn't pass, but `if (typeof module.hot === 'object')` does pass. It's a strange situation.

**Comment by JounQin** [comment]
Ah, it's my pleasure.

I am just using `koa-webpack`, and just want to make it works after upgrading to 2.0, then I find problems, so I must resolve them. 😂

Besides `module.hot` issue wastes me a few hours again to find out that it is not my own problem. 😅

**Comment by shellscape** [comment]
OK we think we have a solution. We think that the HMR Plugin needs to be applied first, before other plugins. Will be testing...

**Comment by shellscape** [comment]
@JounQin we're tracking the update needed for this at https://github.com/webpack/webpack/pull/6296

**Comment by shellscape** [comment]
@JounQin check out this branch to see if it solves the issue on your end. It tests good on my end. https://github.com/webpack-contrib/webpack-hot-client/tree/module-hot-falsy

**Comment by JounQin** [comment]
@shellscape That works to me also.

**Comment by shellscape** [comment]
Awesome. I'll get a patch release up later tonight.

**Comment by shellscape** [comment]
Fixed in #7
